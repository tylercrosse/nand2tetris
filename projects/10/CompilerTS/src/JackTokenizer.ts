export const TokenTypes = {
  KEYWORD: "keyword",
  SYMBOL: "symbol",
  IDENTIFIER: "identifier",
  INT_CONST: "integerConstant",
  STRING_CONST: "stringConstant",
};
export const KeyWordTypes = {
  CLASS: "class",
  METHOD: "method",
  FUNCTION: "function",
  CONSTRUCTOR: "constructor",
  INT: "int",
  BOOLEAN: "boolean",
  CHAR: "char",
  VOID: "void",
  VAR: "var",
  STATIC: "static",
  FIELD: "field",
  LET: "let",
  DO: "do",
  IF: "if",
  ELSE: "else",
  WHILE: "while",
  RETURN: "return",
  TRUE: "true",
  FALSE: "false",
  NULL: "null",
  THIS: "this",
};

/**
 * Tokenizer for Jack language
 */
export default class JackTokenizer {
  commentRegExp = new RegExp(/\/\*(\*(?!\/)|[^*])*\*\/|\/\/.*/g);
  keywordRegExp = new RegExp(
    /class|constructor|function|method|field|static|var|int|char|boolean|void|true|false|null|this|let|do|if|else|while|return/g
  );
  symbolRegExp = new RegExp(
    /\{|\}|\(|\)|\[|\]|\.|\,|\;|\+|\-|\*|\/|\&|\||\<|\>|\=|\~/g
  );
  integerRegExp = new RegExp(/\d+/g);
  stringRegExp = new RegExp(/\"(.*?)\"/g); // needs to use matching groups to remove the quotes
  identifierRegExp = new RegExp(/\w+/g);
  tokenRegexp = new RegExp(
    this.keywordRegExp.source +
      "|" +
      this.symbolRegExp.source +
      "|" +
      this.integerRegExp.source +
      "|" +
      this.stringRegExp.source +
      "|" +
      this.identifierRegExp.source,
    "g"
  );

  tokens: string[] = [];
  currentToken: string;

  constructor(input: string) {
    this.tokens = input.replace(this.commentRegExp, "").match(this.tokenRegexp);
  }

  getCurrentToken() {
    return this.currentToken;
  }

  /**
   * Are there more tokens in the input?
   */
  hasMoreTokens(): boolean {
    return this.tokens.length > 0;
  }

  /**
   * Peeks at the next token.
   * @returns next token
   */
  peekNextToken(): string {
    return this.tokens[0];
  }

  /**
   * Gets the next token from the input and makes it the current token. This method should be called if hasMoreTokens() is true.
   */
  advance(): void {
    if (!this.hasMoreTokens()) return;

    this.currentToken = this.tokens.shift();
  }

  /**
   * Returns the type of the current token.
   */
  tokenType(token: string = this.currentToken) {
    if (this.keywordRegExp.test(token)) {
      this.keywordRegExp.lastIndex = 0; // reset regex
      return TokenTypes.KEYWORD;
    } else if (this.symbolRegExp.test(token)) {
      this.symbolRegExp.lastIndex = 0; // reset regex
      return TokenTypes.SYMBOL;
    } else if (this.integerRegExp.test(token)) {
      this.integerRegExp.lastIndex = 0; // reset regex
      return TokenTypes.INT_CONST;
    } else if (this.stringRegExp.test(token)) {
      this.stringRegExp.lastIndex = 0; // reset regex
      return TokenTypes.STRING_CONST;
    } else {
      return TokenTypes.IDENTIFIER;
    }
  }

  /**
   * Returns the keyword which is the current token. Should be called only when tokenType() is KEYWORD.
   */
  keyWord(): string {
    if (this.tokenType() === TokenTypes.KEYWORD) return this.currentToken;
  }

  /**
   * Returns the character which is the current token. Should be called only when tokenType() is SYMBOL.
   */
  symbol(): string {
    if (this.tokenType() === TokenTypes.SYMBOL) return this.currentToken;
  }

  /**
   * Returns the identifier which is the current token. Should be called only when tokenType() is IDENTIFIER.
   */
  identifier(): string {
    if (this.tokenType() === TokenTypes.IDENTIFIER) return this.currentToken;
  }

  /**
   * Returns the integer value of the current token. Should be called only when tokenType() is INT_CONST.
   */
  intVal(): number {
    if (this.tokenType() === TokenTypes.INT_CONST)
      return parseInt(this.currentToken, 10);
  }

  /**
   * Returns the string value of the current token. Should be called only when tokenType() is STRING_CONST.
   */
  stringVal(): string {
    if (this.tokenType() === TokenTypes.STRING_CONST) return this.currentToken;
  }
}
