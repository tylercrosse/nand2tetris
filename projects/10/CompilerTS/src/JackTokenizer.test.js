import JackTokenizer, { TokenTypes } from "./JackTokenizer";

describe("JackTokenizer", () => {
  const input = `
// This file is part of www.nand2tetris.org
// and the book "The Elements of Computing Systems"
// by Nisan and Schocken, MIT Press.
// File name: projects/10/ArrayTest/Main.jack

// (identical to projects/09/Average/Main.jack)

/** Computes the average of a sequence of integers. */
class Main {
    function void main() {
        var Array a;
        var int length;
        var int i, sum;
  
  let length = Keyboard.readInt("HOW MANY NUMBERS? ");
  let a = Array.new(length);
  let i = 0;
  
  while (i < length) {
      let a[i] = Keyboard.readInt("ENTER THE NEXT NUMBER: ");
      let i = i + 1;
  }
  
  let i = 0;
  let sum = 0;
  
  while (i < length) {
      let sum = sum + a[i];
      let i = i + 1;
  }
  
  do Output.printString("THE AVERAGE IS: ");
  do Output.printInt(sum / length);
  do Output.println();
  
  return;
    }
}
`;

  it("should be able to tokenize a file", () => {
    const tokenizer = new JackTokenizer(input);

    expect(tokenizer.tokens).toHaveLength(140);

    // check some of the output locally
    expect(tokenizer.tokens.slice(0, 7)).toEqual([
      "class",
      "Main",
      "{",
      "function",
      "void",
      "main",
      "(",
    ]);

    // be more exhaustive
    expect(tokenizer.tokens).toMatchSnapshot();
  });

  it("should be able to determine the type of a token", () => {
    const tokenizer = new JackTokenizer(input);

    const types = tokenizer.tokens.map((token) => tokenizer.tokenType(token));

    // check some of the output locally
    expect(types.slice(0, 7)).toEqual([
      TokenTypes.KEYWORD,
      TokenTypes.IDENTIFIER,
      TokenTypes.SYMBOL,
      TokenTypes.KEYWORD,
      TokenTypes.IDENTIFIER,
      TokenTypes.IDENTIFIER,
      TokenTypes.SYMBOL,
    ]);

    // be more exhaustive
    expect(types).toMatchSnapshot();
  });

  it("should be able to match symbols", () => {
    const tokenizer = new JackTokenizer("");
    const symbols = [
      "{",
      "}",
      "(",
      ")",
      "[",
      "]",
      ".",
      ",",
      ";",
      "+",
      "-",
      "*",
      "/",
      "&",
      "|",
      "<",
      ">",
      "=",
      "~",
    ];
    expect(symbols.every((sym) => tokenizer.tokenType(sym))).toEqual(true);
  });
});
