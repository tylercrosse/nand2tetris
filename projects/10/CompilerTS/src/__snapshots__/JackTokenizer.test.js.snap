// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JackTokenizer should be able to determine the type of a token 1`] = `
Array [
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "IDENTIFIER",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "IDENTIFIER",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "STRING_CONST",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "INT_CONST",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "STRING_CONST",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "INT_CONST",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "INT_CONST",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "INT_CONST",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "INT_CONST",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "STRING_CONST",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "KEYWORD",
  "IDENTIFIER",
  "SYMBOL",
  "IDENTIFIER",
  "KEYWORD",
  "SYMBOL",
  "IDENTIFIER",
  "SYMBOL",
]
`;

exports[`JackTokenizer should be able to tokenize a file 1`] = `
Array [
  "class",
  "Main",
  "{",
  "function",
  "void",
  "main",
  "(",
  ")",
  "{",
  "var",
  "Array",
  "a",
  ";",
  "var",
  "int",
  "length",
  ";",
  "var",
  "int",
  "i",
  ",",
  "sum",
  ";",
  "let",
  "length",
  "=",
  "Keyboard",
  ".",
  "readInt",
  "(",
  "\\"HOW MANY NUMBERS? \\"",
  ")",
  ";",
  "let",
  "a",
  "=",
  "Array",
  ".",
  "new",
  "(",
  "length",
  ")",
  ";",
  "let",
  "i",
  "=",
  "0",
  ";",
  "while",
  "(",
  "i",
  "<",
  "length",
  ")",
  "{",
  "let",
  "a",
  "[",
  "i",
  "]",
  "=",
  "Keyboard",
  ".",
  "readInt",
  "(",
  "\\"ENTER THE NEXT NUMBER: \\"",
  ")",
  ";",
  "let",
  "i",
  "=",
  "i",
  "+",
  "1",
  ";",
  "}",
  "let",
  "i",
  "=",
  "0",
  ";",
  "let",
  "sum",
  "=",
  "0",
  ";",
  "while",
  "(",
  "i",
  "<",
  "length",
  ")",
  "{",
  "let",
  "sum",
  "=",
  "sum",
  "+",
  "a",
  "[",
  "i",
  "]",
  ";",
  "let",
  "i",
  "=",
  "i",
  "+",
  "1",
  ";",
  "}",
  "do",
  "Output",
  ".",
  "printString",
  "(",
  "\\"THE AVERAGE IS: \\"",
  ")",
  ";",
  "do",
  "Output",
  ".",
  "printInt",
  "(",
  "sum",
  "/",
  "length",
  ")",
  ";",
  "do",
  "Output",
  ".",
  "println",
  "(",
  ")",
  ";",
  "return",
  ";",
  "}",
  "}",
]
`;
